test/e2e/storage/in_tree_volumes.go

```diff
@@ -0,0 +1,228 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	"k8s.io/api/core/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/testsuites"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+// List of testDrivers to be executed in below loop
+var testDrivers = []func() drivers.TestDriver{
+	drivers.InitNFSDriver,
+	drivers.InitGlusterFSDriver,
+	drivers.InitISCSIDriver,
+	drivers.InitRbdDriver,
+	drivers.InitCephFSDriver,
+	drivers.InitHostpathDriver,
+	drivers.InitHostpathSymlinkDriver,
+	drivers.InitEmptydirDriver,
+	drivers.InitCinderDriver,
+	drivers.InitGceDriver,
+	drivers.InitVSphereDriver,
+	drivers.InitAzureDriver,
+	drivers.InitAwsDriver,
+}
+
+// List of testSuites to be executed in below loop
+var testSuites = []func() testsuites.TestSuite{
+	testsuites.InitVolumesTestSuite,
+	testsuites.InitVolumeIOTestSuite,
+	testsuites.InitVolumeModeTestSuite,
+	testsuites.InitSubPathTestSuite,
+}
+
+// This executes testSuites for in-tree volumes.
+// It is done by following steps:
+// Loop over testDrivers:
+//   Loop over testSuites:
+//     - Get testPatterns from testSuites
+//     Loop over testPatterns:
+//       - Get test function to be executed
+//       BeforeEach test, prepare test by following steps:
+//          - Skip unsupported tests to avoid unnecessary resource initialization
+//          - Create Driver
+// 			- Create Volume for test unless dynamicPV test
+//          - Initialize test resource for testSuites
+//          - Setup test resource for driver and testpattern
+//          - Create test input to be passed to test function
+//       - Execute tests by calling test function with test input
+//       AfterEach test, cleanup everything by following steps:
+//          - Cleanup test resource
+// 			- Delete Volume for test unless dynamicPV test
+//          - Cleanup Driver
+var _ = utils.SIGDescribe("In-tree Volumes", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		ns     *v1.Namespace
+		config framework.VolumeTestConfig
+
+		drv drivers.TestDriver
+		ts  testsuites.TestSuite
+		tr  testsuites.TestResource
+		ti  testsuites.TestInput
+		vt  testpatterns.TestVolType
+		tp  testpatterns.TestPattern
+
+		needsCleanup bool
+	)
+
+	BeforeEach(func() {
+		ns = f.Namespace
+		config = framework.VolumeTestConfig{
+			Namespace: ns.Name,
+			Prefix:    "volume",
+		}
+	})
+
+	for _, initDriver := range testDrivers {
+		curDriver := initDriver()
+
+		for _, testSuiteInit := range testSuites {
```

Comment:
Can this be enclosed in a function called something like "runTestSuite"?  I want to separate out the driver setup from the test cases to support the scenario where the driver may have already been preinstalled or setup.